File: C/1654.c
#include <stdio.h>  // 랜선 자르기
#include <stdlib.h>

int arr[10001] = { 0, };

long long binary_search(int max, int k, int n)
{
	long long left = 1, right = max, cut_max = 0;
	long long mid;
	while (left <= right)
	{
		mid = (left + right) / 2;

		int cnt = 0;
		for (int i = 0; i < k; i++)
		{
			cnt += arr[i] / mid;
		}

		if (cnt >= n)
		{
			if (cut_max < mid)
				cut_max = mid;
			left = mid + 1;
		}
		else
		{
			right = mid - 1;
		}
	}

	return cut_max;
}


int main()
{
	int k, n;
	scanf("%d %d", &k, &n);

	int max = 0;
	for (int i = 0; i < k; i++)
	{
		scanf("%d", &arr[i]);
		if (arr[i] > max)
			max = arr[i];
	}

	printf("%lld", binary_search(max, k, n));

	return 0;
}

File: JAVA/src/_1654.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

// 랜선 자르기
public class _1654 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        String[] input = br.readLine().split(" ");
        int K = Integer.parseInt(input[0]);
        int N = Integer.parseInt(input[1]);

        long[] arr = new long[K];
        for (int i = 0; i < K; i++) {
            arr[i] = Long.parseLong(br.readLine());
        }

        Arrays.sort(arr);

        System.out.println(BinarySearch(arr, N, K));
    }

    public static long BinarySearch(long[] arr, int n, int k) {
        long left = 1;
        long right = arr[k - 1];
        long ans = 1;
        while (left <= right) {
            long mid = (left + right) / 2;

            long cnt = 0;
            for (int i = 0; i < k; i++) {
                cnt += arr[i] / mid;
            }

            if (cnt >= n) {
                ans = Math.max(ans, mid);
                left = mid + 1;
            }
            else {
                right = mid - 1;
            }
        }

        return ans;
    }
}
